emu_sv/sv_backend_impl.py

```diff
@@ -0,0 +1,247 @@
+            )
+
+    def _save_statistics(self, step_idx: int) -> None:
+        norm_time = self.target_times[step_idx + 1] / self.target_times[-1]
+        self.statistics.data.append(time.time() - self.time)
+        self.statistics(
+            self._config,
+            norm_time,
+            self.state,
+            self._current_H,  # type: ignore[arg-type]
+            self.results,
+        )
+        self.time = time.time()
+        self._current_H = None
+
+    def _run(self) -> Results:
+        for step in range(self.nsteps):
+            self.step(step)
+
+        return self.results
+
+
+class SVBackendImpl(BaseSVBackendImpl):
+
+    def __init__(self, config: SVConfig, pulser_data: PulserData):
+        """
+        For running sequences without noise. The state will evolve accoring
+        to e^(-iH t)
+
+        Args:
+            config: The configuration for the emulator.
+            pulser_data: The data for the sequence to be emulated.
+        """
+        super().__init__(config, pulser_data)
+
+        self.state: StateVector = (
+            StateVector.make(self.nqubits, gpu=self._config.gpu)
+            if self._config.initial_state is None
+            else StateVector(
+                self._config.initial_state.vector.clone(),
+                gpu=self._config.gpu,
+            )
+        )
+
+        self.stepper = (
+            EvolveStateVector.apply
+            if self.state.vector.requires_grad
+            else EvolveStateVector.evolve
+        )
+
+    def _evolve_step(self, dt: float, step_idx: int) -> None:
+        self.state.vector, self._current_H = self.stepper(
+            dt * _TIME_CONVERSION_COEFF,
+            self.omega[step_idx],
+            self.delta[step_idx],
+            self.phi[step_idx],
+            self._pulser_data.full_interaction_matrix,
+            self.state.vector,
+            self._config.krylov_tolerance,
+        )
+
+
+class NoisySVBackendImpl(BaseSVBackendImpl):
+
+    def __init__(self, config: SVConfig, pulser_data: PulserData):
+        """
+        Initializes the NoisySVBackendImpl, master equation version.
+        This class handles the Lindblad operators and
+        solves the Lindblad master equation
+
+        Args:
+            config: The configuration for the emulator.
+            pulser_data: The data for the sequence to be emulated.
+        """
+
+        super().__init__(config, pulser_data)
+
+        self.pulser_lindblads = pulser_data.lindblad_ops
+
+        self.state: DensityMatrix = (
+            DensityMatrix.make(self.nqubits, gpu=self._config.gpu)
+            if self._config.initial_state is None
+            else DensityMatrix(
+                self._config.initial_state.matrix.clone(), gpu=self._config.gpu
+            )
+        )
+
+        self.stepper = EvolveDensityMatrix.evolve
+
+    def _evolve_step(self, dt: float, step_idx: int) -> None:
+        self.state.matrix, self._current_H = self.stepper(
```

Comment:
what I meant with removing self.stepper  is directly calling `EvolveDensityMatrix.evolve` here